internal/internal_worker_base.go

```diff
@@ -401,3 +413,42 @@ func (bw *baseWorker) Stop() {
 	}
 	return
 }
+
+func (bw *baseWorker) emitHardwareUsage() {
+	defer func() {
+		if p := recover(); p != nil {
+			bw.metricsScope.Counter(metrics.WorkerPanicCounter).Inc(1)
+			topLine := fmt.Sprintf("base worker for %s [panic]:", bw.options.workerType)
+			st := getStackTraceRaw(topLine, 7, 0)
+			bw.logger.Error("Unhandled panic in hardware emitting.",
+				zap.String(tagPanicError, fmt.Sprintf("%v", p)),
+				zap.String(tagPanicStack, st))
+		}
+	}()
+	defer bw.shutdownWG.Done()
+	ticker := time.NewTicker(hardwareMetricsCollectInterval)
+	for {
+		select {
+		case <-bw.shutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			host := bw.options.host
+			scope := bw.metricsScope.Tagged(map[string]string{clientHostTag: host})
+
+			cpuPercent, _ := cpu.Percent(0, false)
+			cpuCores, _ := cpu.Counts(false)
+
+			scope.Gauge(metrics.NumCPUCores).Update(float64(cpuCores))
+			scope.Gauge(metrics.CPUPercentage).Update(cpuPercent[0])
```

Comment:
we might still have a nil pointer if cpu.Percent(0, false) returns error. Either need to handle error or do a length check.